src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java

```diff
@@ -196,61 +198,115 @@ private LiKafkaProducerImpl(LiKafkaProducerConfig configs,
-        assert (wrappedRecord.size() == 1);
-        future = _producer.send(wrappedRecord.get(0), errorLoggingCallback);
+        new ErrorLoggingCallback<>(auditToken, topic, timestamp, sizeInBytes, _auditor, callback);
+
+      if (useLargeMessageProcessing) {
+        errorLoggingCallback = new LargeMessageCallback(xRecords.size(), errorLoggingCallback);
+      }
+      Future<RecordMetadata> future = null;
+      for (ExtensibleProducerRecord<byte[], byte[]> segmentRecord : xRecords) {
+        ProducerRecord<byte[], byte[]> segmentProducerRecord = serializeWithHeaders(segmentRecord, _headerSerializer);
+        future = _producer.send(segmentProducerRecord, errorLoggingCallback);
       }
       return future;
-    } catch (Throwable t) {
+    } catch (Exception t) {
       _auditor.record(_auditor.auditToken(producerRecord.key(), producerRecord.value()), producerRecord.topic(),
-                      producerRecord.timestamp(), 1L, 0L, AuditType.FAILURE);
+          producerRecord.timestamp(), 1L, 0L, AuditType.FAILURE);
       throw new KafkaException(t);
     } finally {
       _numThreadsInSend.decrementAndGet();
     }
   }
 
+  private Future<RecordMetadata> sendNullValue(ProducerRecord<K, V> producerRecord, Callback callback, String topic,
+    K key, V value, Long timestamp, byte[] serializedValue, byte[] serializedKey, int sizeInBytes, Object auditToken) {
+    Future<RecordMetadata> future; // We wrap the user callback for error logging and auditing purpose.
+    Callback errorLoggingCallback =
+      new ErrorLoggingCallback<>(auditToken, topic, timestamp, sizeInBytes, _auditor, callback);
+    _auditor.record(auditToken, topic, timestamp, 1L, (long) sizeInBytes, AuditType.ATTEMPT);
+    ProducerRecord<byte[], byte[]> headerlessByteRecord =
+      new ProducerRecord<>(producerRecord.topic(), producerRecord.partition(), producerRecord.timestamp(),
+        serializedKey, serializedValue);
+    future = _producer.send(headerlessByteRecord, errorLoggingCallback);
+    return future;
+  }
+
+  /**
+   * This is public for testing.
+   * This assumes that the value is not a large value that would be segmented by large message support.
+   */
+  public static ProducerRecord<byte[], byte[]> serializeWithHeaders(ExtensibleProducerRecord<byte[], byte[]> xRecord,
+    HeaderSerializer headerParser) {
+
+    if (!xRecord.hasHeaders()) {
+      return new ProducerRecord<byte[], byte[]>(xRecord.topic(), xRecord.partition(), xRecord.timestamp(), xRecord.key(), xRecord.value());
+    }
+
+    int headersSize = headerParser.serializedHeaderSize(xRecord.headers());
+    int serializedValueSize = (xRecord.value() == null ? 0 : xRecord.value().length) + headersSize;
+
+    ByteBuffer valueWithHeaders = ByteBuffer.allocate(serializedValueSize);
+    headerParser.serializeHeader(valueWithHeaders, xRecord.headers(), xRecord.value() == null);
```

Comment:
This method can't be moved to HeaderUtils because it is dependent on the header serializer used.  I'm not sure how value can be a ByteBuffer since to avoid copying the ByteBuffer the header is seriailzed into needs to be serializedValue.length + serializedHeader.length.